core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -1952,9 +1952,8 @@ private static RexNode isDistinctFromInternal(
         rexBuilder.makeCall(nullOp, x),
 
         // else return x compared to y
-        rexBuilder.makeCall(eqOp,
-            rexBuilder.makeNotNull(x),
-            rexBuilder.makeNotNull(y))
+        rexBuilder.makeCall(SqlStdOperatorTable.IS_TRUE,
+            rexBuilder.makeCall(eqOp, x, y))
```

Comment:
I wonder if the following should be used instead:

```
    return rexBuilder.makeCall(SqlStdOperatorTable.OR,
        rexBuilder.makeCall(eqOp,
          rexBuilder.makeCall(SqlStdOperatorTable.IS_NULL, x),
          rexBuilder.makeCall(SqlStdOperatorTable.IS_NULL, y)),
        rexBuilder.makeCall(SqlStdOperatorTable.IS_TRUE,
          rexBuilder.makeCall(eqOp, x, y)));
```

It would make `nullOp` obsolete.

Reply:
Yes, it could be done - there is also is not distinct; so there will be 2 blocks of call blocks.

While I was working on adding better expressions for them - I've checked some plan changes...and it seems we are not taking advantage of the following simplification opportunities:

* complementary IS predicates under ANDs:
  ```
    AND(
      IS NULL($12), 
      IS NULL($3),
      IS NOT NULL($3)
    ),
  ```
* this might be too much but some notes about:
  ```
    AND(
      IS TRUE(=($3, $12)),
      IS NOT NULL($3),
      IS NOT NULL($12)
    )
  ```
  * `IS TRUE(=($3, $12))` implies `IS NOT NULL($3)`; basically a comparision under an `IS TRUE` implies that the operands are `NOT NULL`...

with the above 2 rules; `RexSimplify` would been able to rewrite to the new, better expressions by itself :)